rules/prefer-error-cause.js

```diff
@@ -0,0 +1,364 @@
+'use strict';
+const {isCommaToken} = require('eslint-utils');
+const _ = require('lodash');
+const {matches, methodCallSelector} = require('./selectors/index.js');
+const {appendArgument, replaceReferenceIdentifier} = require('./fix/index.js');
+const getReferences = require('./utils/get-references.js');
+
+const ERROR = 'error';
+const SUGGESTION = 'suggestion';
+
+const messages = {
+	[ERROR]: 'Prefer specifying an old error as `cause` option where rethrowing the error.',
+	[SUGGESTION]: 'Specify `cause` option to Error\'s constructor.',
+};
+
+const catchClauseSelector = matches([
+	[
+		'CatchClause',
+		':has(ThrowStatement)',
+	].join(''),
+]);
+
+const promiseThenCatchSelector = matches([
+	[
+		matches([
+			methodCallSelector({method: 'then', argumentsLength: 2}),
+			methodCallSelector({method: 'catch', argumentsLength: 1}),
+		]),
+		':has(ThrowStatement)',
+	].join(''),
+]);
+
+const peekTop = array => array[array.length - 1];
+
+const insertProperty = (fixer, node, text, sourceCode) => {
+	const [penultimateToken, lastToken] = sourceCode.getLastTokens(node, 2);
+	if (node.properties.length > 0) {
+		text = isCommaToken(penultimateToken) ? `${text}` : `, ${text}`;
+	}
+
+	return fixer.insertTextBefore(lastToken, text);
+};
+
+const generateCauseInspector = inspectProperty => objectExpression => {
+	if (!objectExpression || objectExpression.type !== 'ObjectExpression') {
+		return false;
+	}
+
+	return objectExpression.properties.some(property =>
+		!property.computed && property.key.name === 'cause' && inspectProperty(property),
+	);
+};
+
+const reportCannotBeFixed = (node, context) => {
+	context.report({
+		node,
+		messageId: ERROR,
+	});
+};
+
+const fixerUtils = {
+	insertCauseArgument({fixer, node, sourceCode, value}) {
+		return appendArgument(
+			fixer,
+			node,
+			`{cause: ${value}}`,
+			sourceCode,
+		);
+	},
+
+	insertCauseProperty({fixer, node, sourceCode, value}) {
+		return insertProperty(fixer, node, `cause: ${value}`, sourceCode);
+	},
+
+	renameObjectProperty({fixer, node, key, value}) {
+		if (!node || node.type !== 'ObjectExpression') {
+			return false;
+		}
+
+		const identifier = node.properties.find(property => {
+			if (property.key.name === key) {
+				return property;
+			}
+
+			return undefined;
+		});
+
+		if (!identifier) {
+			throw new Error(`objectExpression doesn't have ${key} property`);
+		}
+
+		return replaceReferenceIdentifier(identifier.value, value, fixer);
+	},
+
+	handleEmptyArgument({fixer, node, sourceCode, errorArgumentIdentifier}) {
```

Comment:
Why not split this into two functions, `insertErrorCatchClauseParameter` and `insertFunctionParameter`, if I didn't get this wrong.